### ISSUE TITLE
Workflow Controller error when using withItems and hooks

### ISSUE THREAD

**Issue by calvinwiebe** [argoproj/argo-workflows#12279]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

When using `withItems` on a DAG task in combination with the same task having `hooks`, the overall DAG fails due to a `Workflow Controller error` with the specific message of `runtime error: index out of range [1] with length 1`. 

I expected `withItems` and `hooks` to work together. 

### Version

v3.5.2

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
  kind: Workflow
  metadata:
    namespace: argo
    generateName: with-items-dag-error-
  spec:
    podMetadata:
      labels:
        parent-workflow:  '{{workflow.name}}'
    podGC:
      strategy: OnPodCompletion
    templates:
      - name: hook-template
        nodeSelector:
          metrics.aurora.workload.type.simulation: 'true'
          metrics.aurora.workload.execution-mode.job: 'true'
          metrics.aurora.workload.machine-class.s4: 'true'
        metadata:
          annotations:
            user_created_task: 'false'
        inputs:
          parameters:
            - name: 'foo'
        script:
          image: >-
            alpine:3.17.3
          command:
            - sh
          source: >-
            echo "I am a hook {{inputs.parameters.foo}}"
      - name: run-sim-template
        nodeSelector:
          metrics.aurora.workload.type.simulation: 'true'
          metrics.aurora.workload.execution-mode.job: 'true'
          metrics.aurora.workload.machine-class.s4: 'true'
        metadata:
          annotations:
            user_created_task: 'true'
        inputs:
          parameters:
            - name: fanoutValue
        script:
          image: >-
            alpine:3.17.3
          command:
            - sh
          source: >-
            echo $FANOUT_VALUE
          env:
            - name: FANOUT_VALUE
              value: '{{inputs.parameters.fanoutValue}}'
      - name: run-sim-steps-template
        inputs:
          parameters:
            - name: fanoutValue
        steps:
          - - name: run-sim
              arguments:
                parameters:
                  - name: fanoutValue
                    value: '{{inputs.parameters.fanoutValue}}'
              template: run-sim-template
      - name: main
        dag:
          tasks:
            - name: run-sim
              arguments:
                parameters:
                  - name: fanoutValue
                    value: '{{item}}'
              template: run-sim-steps-template
              withItems:
                - 1
                - 2
                - 3
              hooks:
                exit:
                  template: hook-template
                  arguments:
                    parameters:
                      - name: foo
                        value: bar
                running:
                  template: hook-template
                  arguments:
                    parameters:
                      - name: foo
                        value: bar
                  expression: tasks["run-sim"].status == "Running"
    artifactGC:
      strategy: OnWorkflowCompletion
      serviceAccountName: argo-executor
      forceFinalizerRemoval: true
    tolerations:
      - key: kubernetes.azure.com/scalesetpriority
        operator: Equal
        value: spot
        effect: NoSchedule
    entrypoint: main
```

### Logs from the workflow controller

```text
time="2023-11-22T14:56:16.640Z" level=error msg="Recovered from panic" namespace=fanout-test-oo94j r="runtime error: index out of range [1] with length 1" stack="goroutine 319 [running]:\nruntime/debug.Stack()\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x65\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2()\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:193 +0xbc\npanic({0x2104420, 0xc012e499c8})\n\t/usr/local/go/src/runtime/panic.go:890 +0x263\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.parseLoopIndex({0xc022f826f0?, 0x4160b0?})\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2878 +0x146\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.loopNodes.Less({0xc0019be000, 0x5, 0x4fcdbf?}, 0xc03a5f0e72?, 0x2)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2888 +0x4d\nsort.insertionSort({0x259f8d0, 0xc002253140}, 0x0, 0x5)\n\t/usr/local/go/src/sort/zsortinterface.go:12 +0xb1\nsort.pdqsort({0x259f8d0, 0xc002253140}, 0x18?, 0x1ebc100?, 0x501601?)\n\t/usr/local/go/src/sort/zsortinterface.go:73 +0x2f7\nsort.Sort({0x259f8d0, 0xc002253140})\n\t/usr/local/go/src/sort/sort.go:48 +0x5d\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).processAggregateNodeOutputs(0x21317a0?, 0xc0185c7c00, {0xc01f456ed0, 0xd}, {0xc0019be000?, 0x5, 0xc00198c840?})\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2904 +0x8e\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAG(0xc000724580, {0x25a0b48, 0xc00005c018}, {0xc01be24428, 0x3}, 0xc015a86000, {0xc009fc0c64, 0x6}, 0xc03ed7f680, {0x25a42b8, ...}, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:303 +0xefa\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc000724580, {0x25a0b48, 0xc00005c018}, {0xc01be24428, 0x3}, {0x25a42b8, 0xc001919bc0?}, 0xc001911590?, {{0x0, 0x0, ...}, ...}, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2004 +0x20e5\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate(0xc000724580, {0x25a0b48, 0xc00005c018})\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:349 +0x172d\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem(0xc000494500, {0x25a0b48, 0xc00005c018})\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:768 +0x759\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc039af76a0?)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:690 +0x9e\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.24.3/pkg/util/wait/wait.go:155 +0x3e\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x257ea00, 0xc039d8c900}, 0x1, 0xc00053cfc0)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.24.3/pkg/util/wait/wait.go:156 +0xb6\nk8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.24.3/pkg/util/wait/wait.go:133 +0x89\nk8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0x0?, 0x0?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.24.3/pkg/util/wait/wait.go:90 +0x25\ncreated by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).Run\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:317 +0x1a45\n" workflow=job
time="2023-11-22T14:56:16.640Z" level=info msg="Updated message  -> runtime error: index out of range [1] with length 1" namespace=fanout-test-oo94j workflow=job
```

### Logs from in your workflow's wait container

```text
N/A
```

**Comment by calvinwiebe** [comment]
[Thread confirming issue in slack](https://cloud-native.slack.com/archives/C01QW9QSSSK/p1700677770029819)

tl;dr - `parseLoopIndex` in [operator.go](https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/operator.go#L3048-L3075) gets fed both the withItems _and_ the hook nodes. The hook nodes fail the DisplayName parse.

**Comment by Joibel** [comment]
This is caused by `processAggregateNodeOutputs` being called when not all children are withItem/withParam children, so attempting to parse DisplayName fails.

Suggested fix is to filter children to those which are withItem/Param children in `processAggregateNodeOutputs` before working on them.
